docs/transparent-proxy.md

```diff
@@ -0,0 +1,268 @@
+TABLE_ID=100
+
+# Create a custom routing table for pyrdp traffic
+echo "$TABLE_ID    pyrdp" >> /etc/iproute2/rt_tables
+
+# Intercept RDP traffic destined for the server (1)
+iptables -t mangle -I PREROUTING -p tcp -d 10.13.37.111 --dport 3389 \
+  -j TPROXY --tproxy-mark 0x1/0x1 --on-port 3389 --on-ip 127.0.0.1
+
+# Mark RDP return traffic intended for clients (2)
+iptables -t mangle -A PREROUTING \
+    -s $SERVER_IP \
+    -m tcp -p tcp --sport 3389 \
+    -j MARK --set-mark $MARK
+
+# Set route lookup to the pyrdp table for marked packets.
+ip rule add fwmark $MARK lookup $TABLE_ID
+
+# Add a custom route that redirects traffic intended for the outside world to loopback
+# So that server-client traffic passes through PyRDP
+# This table will only ever be used by RDP so it should not be problematic
+ip route add local default dev lo table $TABLE_ID
+
+# If you want the interception to happen at the Layer 2
+# Make sure that both your interfaces are bridged together
+
+# WARNING: if you have other important firewall rules, make sure to test the impact of this
+modprobe br_netfilter
+echo 1 > /proc/sys/net/bridge/bridge-nf-call-iptables
+
+# Target DROP in brouting means out of L2 go route in L3
+ebtables -t broute -A BROUTING -i enp0s3 -p ipv4 --ip-dst $SERVER_IP --ip-proto tcp --ip-dport 3389 -j redirect --redirect-target DROP
+ebtables -t broute -A BROUTING -i enp0s8 -p ipv4 --ip-src $SERVER_IP --ip-proto tcp --ip-source-port 3389 -j redirect --redirect-target DROP
+# recent linux like debian buster
+#ebtables-legacy -t broute -A BROUTING -i enp0s8 -p ipv4 --ip-src $SERVER_IP --ip-proto tcp --ip-source-port 3389 -j redirect --redirect-target DROP
+
+# Disable return path filtering
+echo 0 > /proc/sys/net/ipv4/conf/default/rp_filter
+echo 0 > /proc/sys/net/ipv4/conf/all/rp_filter
+echo 0 > /proc/sys/net/ipv4/conf/enp0s3/rp_filter
+echo 0 > /proc/sys/net/ipv4/conf/enp0s8/rp_filter
+
+# Make interception work without an IP assigned to br0
+# WARNING: unclear if required on not
+echo 1 > /proc/sys/net/ipv4/conf/br0/route_localnet
+echo 1 > /proc/sys/net/ipv4/conf/enp0s3/route_localnet
+
+# Mark socket Packets
+# WARNING: Unclear if required or not but likely good for performance so it will likely stay
+iptables -t mangle -N DIVERT
+iptables -t mangle -A DIVERT -j MARK --set-mark 1
+iptables -t mangle -A DIVERT -j ACCEPT
+iptables -t mangle -I PREROUTING -p tcp -m socket -j DIVERT
+```
+
+In order for L3 traffic to be properly routed between the client and servers, there are two options.
+
+### Option 1: Add an Interface with a valid IP in the networks of interests on the Bridge
+
+This is important to allow ARP resolution to work properly (when the MITM unavoidably needs to know who has the
+server/client IP)
+
+In the example setup, doing something like below is enough to make everything work because when the MITM needs the L2
+address to reply to, it has the means to look it up by itself.
+
+```
+ip addr add 10.13.37.11/24 dev br0
```

Comment:
Without the default gateway you can't reply to non-local RDP connections. 192.168.0.201 isn't for me, I don't have a route, I don't have a default route, good bye.

Reply:
Fixed